File: examples/cpp/my-app/src/database/UserRepository.cpp
#include "database/UserRepository.h"
#include "utils/Logger.h"

namespace database {

UserRepository::UserRepository() : db(Database::getInstance()) {}

std::shared_ptr<models::User> UserRepository::create(const std::string& name, const std::string& email) {
    db.execute("INSERT INTO users (name, email) VALUES ('" + name + "', '" + email + "')");
    return std::make_shared<models::User>(name, email);
}

std::shared_ptr<models::User> UserRepository::findById(int id) {
    db.execute("SELECT * FROM users WHERE id = " + std::to_string(id));
    return nullptr; // Simplified
}

std::shared_ptr<models::User> UserRepository::findByEmail(const std::string& email) {
    db.execute("SELECT * FROM users WHERE email = '" + email + "'");
    return std::make_shared<models::User>("", email); // Simplified
}

std::vector<models::User> UserRepository::findAll() {
    db.execute("SELECT * FROM users");
    return {};
}

void UserRepository::update(int id, const std::string& name, const std::string& email) {
    db.execute("UPDATE users SET name = '" + name + "', email = '" + email + "' WHERE id = " + std::to_string(id));
}

void UserRepository::deleteById(int id) {
    db.execute("DELETE FROM users WHERE id = " + std::to_string(id));
}

void UserRepository::save(const models::User& user) {
    db.execute("UPDATE users SET name = '" + user.getName() + "', email = '" + user.getEmail() + "'");
}

} // namespace database

File: examples/python/my-app/src/database/user_repository.py
"""User repository for database operations."""

from typing import List, Optional
from models.user import User
from database.database_manager import DatabaseManager
from utils.logger import Logger

class UserRepository:
    """Repository for User entity database operations."""
    
    def __init__(self):
        self._db = DatabaseManager.get_instance()
        self._logger = Logger.get_instance()
    
    def find_by_id(self, user_id: int) -> Optional[User]:
        """Find user by ID."""
        self._logger.debug(f"Finding user with ID: {user_id}")
        query = "SELECT * FROM users WHERE id = :id"
        result = self._db.execute_query(query, {"id": user_id})
        # Mock return
        return None
    
    def find_by_email(self, email: str) -> Optional[User]:
        """Find user by email."""
        self._logger.debug(f"Finding user with email: {email}")
        query = "SELECT * FROM users WHERE email = :email"
        result = self._db.execute_query(query, {"email": email})
        return None
    
    def find_all(self) -> List[User]:
        """Find all users."""
        self._logger.debug("Finding all users")
        query = "SELECT * FROM users"
        result = self._db.execute_query(query)
        return []
    
    def save(self, user: User) -> User:
        """Save user to database."""
        self._logger.info(f"Saving user: {user.username}")
        query = "INSERT INTO users (...) VALUES (...)"
        result = self._db.execute_query(query)
        return user
    
    def update(self, user: User) -> User:
        """Update existing user."""
        self._logger.info(f"Updating user: {user.id}")
        query = "UPDATE users SET ... WHERE id = :id"
        result = self._db.execute_query(query, {"id": user.id})
        return user
    
    def delete(self, user_id: int) -> bool:
        """Delete user by ID."""
        self._logger.warning(f"Deleting user: {user_id}")
        query = "DELETE FROM users WHERE id = :id"
        result = self._db.execute_query(query, {"id": user_id})
        return True
